url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/XMLDecoder.java

```diff
@@ -0,0 +1,705 @@
+            elem.id = attributes.getValue("id");
+            elem.idref = attributes.getValue("idref");
+            elem.attributes = attributes;
+            elem.target = tagName;
+            readObjs.push(elem);
+        }
+
+        @Override
+        public void endElement(String uri, String localName, String qName)
+                throws SAXException {
+            if (!inJavaElem) {
+                return;
+            }
+            if ("java".equals(qName)) { //$NON-NLS-1$
+                inJavaElem = false;
+                return;
+            }
+            // find the elem to close
+            Elem toClose = latestUnclosedElem();
+            if ("string".equals(toClose.target)) {
+                StringBuilder sb = new StringBuilder();
+                for (int index = readObjs.size() - 1; index >= 0; index--) {
+                    Elem elem = (Elem) readObjs.get(index);
+                    if (toClose == elem) {
+                        break;
+                    }
+                    if ("char".equals(elem.target)) {
+                        sb.insert(0, elem.methodName);
+                    }
+                }
+                toClose.methodName = toClose.methodName != null ? toClose.methodName
+                        + sb.toString()
+                        : sb.toString();
+            }
+            // make sure it is executed
+            execute(toClose);
+            // set to closed
+            toClose.isClosed = true;
+            // pop it and its children
+            while (readObjs.pop() != toClose) {
+                //
+            }
+
+            if (toClose.isExpression) {
+                // push back expression
+                readObjs.push(toClose);
+            }
+        }
+
+        private Elem latestUnclosedElem() {
+            for (int i = readObjs.size() - 1; i >= 0; i--) {
+                Elem elem = readObjs.get(i);
+                if (!elem.isClosed) {
+                    return elem;
+                }
+            }
+            return null;
+        }
+
+        private Object execute(Elem elem) {
+            if (elem.isExecuted) {
+                return elem.result;
+            }
+
+            // execute to obtain result
+            try {
+                if (elem.idref != null) {
+                    elem.result = idObjMap.get(elem.idref);
+                } else if (elem.isBasicType) {
+                    elem.result = executeBasic(elem);
+                } else {
+                    elem.result = executeCommon(elem);
+                }
+            } catch (Exception e) {
+                listener.exceptionThrown(e);
+            }
+
+            // track id
+            if (elem.id != null) {
+                idObjMap.put(elem.id, elem.result);
+            }
+
+            elem.isExecuted = true;
+            return elem.result;
+        }
+
+        @SuppressWarnings("nls")
+        private Object executeCommon(Elem elem) throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 361-385 into a separate method.